### PR TITLE
Exciton

### PR DESCRIPTION
**Pull request by lanetolley**
## Description
Add in ExcitonDriver class to Wptherml main branch

## Todos
Notable points that this PR has either accomplished or will accomplish.
  - [x] SpinBosonDriver class implements Tavis-Cummings model implemented - can support N spins and 1 M-level boson
  - [x] Unit tests for SpinBosonDriver class implemented
  - [x] High-level test for TC model in the 1-spin and 2-level boson case implemented and passing against results from QuTip 
  - [x] Dissipative dynamics implemented and tested against 1-spin 2-level boson case from QuTip with gamma = 5e-6 atomic units
  - [x] Need to implement tests for more than 2 spins, can test against @PeytonRoden's code and/or QuTip
  - [x] Need to implement tests for 1 spin with more than 2 boson levels, can use QuTip
  - [ ] Need to implement tests for dynamics for more 2 or more spins and with additional types of dissipation (dephasing, dissipation in the boson case) can use Peyton's code for this
  - [ ] Current coupling uses the RWA and is parameterized only by a single constant g, should generalize to include (optional) counter rotating terms and the more flexible coupling of the field vector dotted into the dipole operator
  - [ ] Current model excludes the dipole self energy, I will implement this

## Questions
- []

## Status
- [Pending review] Ready to go

### CODE REVIEW

File: wptherml/exciton.py [hunk 1]
```diff
@@ -0,0 +1,540 @@
...
+    def parse_input(self, args):
+        if "exciton_energy" in args:
+            self.exciton_energy = args["exciton_energy"]
+        else:
+            self.exciton_energy = 0.5
+        if "number_of_monomers" in args:
+            self.number_of_monomers = args["number_of_monomers"]
+        else:
+            self.number_of_monomers = 2
+        if "displacement_between_monomers" in args:
```
Comment (FoleyLab) [inline wptherml/exciton.py]: I think there is a unit mismatch between the displacement_between_monomers here and the default displacements on line 72 and 77.  I suspect the vert_ and horz_ and diag_ displacements are in atomic units, can you confirm?  With vert_, horz_, diag_, there probably is not need for the original displacement_between_monomers anymore, so we should remove it.

File: wptherml/exciton.py [hunk 2]
```diff
@@ -0,0 +1,540 @@
...
+
+
+class ExcitonDriver(SpectrumDriver):
+    """A class for computing the dynamics and spectra of a system modelled by the Frenkel Exciton Hamiltonian
+
+    Attributes
+    ----------
+    radius : float
+        the radius of the sphere
+
```
Comment (FoleyLab) [inline wptherml/exciton.py:31]: The attributes should include the attributes that are assigned values from the parse_input() and the __init__ methods; it would be helpful to clarify the units that the attributes are expressed in when appropriate

Here are a few for example:

Atributes
---------
exciton_hamiltonian : numpy array of floats
    the Frenkel exciton Hamiltonian in atomic units

exciton_energy : float
    the energy of the monomer exciton in atomic units

number_of_monomers : int
    the number of monomers in the simulation

File: wptherml/exciton.py [hunk 3]
```diff
@@ -0,0 +1,540 @@
...
+        if "transition_dipole_moment" in args:
+            self.transition_dipole_moment = args["transition_dipole_moment"]
+        else:
+            self.transition_dipole_moment = np.array([0, 0, 1])
+        if "refractive_index" in args:
+            self.refractive_index = args["refractive_index"]
+        else:
+            self.refractive_index = 1
+        
+        if "vert_displacement_between_monomers" in args:
```
Comment (FoleyLab) [inline wptherml/exciton.py]: Please use the full words in snake case to avoid confusion for both the keys and the attribute names:
"horz_displacement_between_monomers" -> "horizontal_displacement_between_monomers" 
self.horz_displacement_between_monomers -> self.horizontal_displacement_between_monoers

File: wptherml/exciton.py [hunk 4]
```diff
@@ -0,0 +1,594 @@
...
+        #############################################################
+        # going to change this to aggregate_shape 
+        #if "number_of_monomers" in args:
+        #    self.number_of_monomers = args["number_of_monomers"]
+        #else:
+        #    self.number_of_monomers = 2
+        #############################################################
+
+        if "aggregate_shape" in args:
+            self.aggregate_shape = args["aggregate_shape"]
```
Comment (FoleyLab) [inline wptherml/exciton.py:77]: @lanetolley I added this input flag which is meant to be a tuple specifying the shape of the aggregate in terms of number of repeats of the monomer along the x, y, and z axis. For example, a 2D aggregate with 4 total monomers, 2 monomers along the x-axis and 2 along the y-axis, would be specified like (2, 2, 1).  A 3D aggregate with 8 total monomers, 2 along x, 2 along y, 2 along z, would be (2, 2, 2), etc.

File: wptherml/exciton.py [hunk 5]
```diff
@@ -0,0 +1,594 @@
...
+        
+        # user might accidentally make one element of the tuple zero
+        if self.aggregate_shape[0] == 0 or self.aggregate_shape[1] == 0 or self.aggregate_shape[2] == 0:
+            print(" Invalid shape!  All directions must have at least 1 layer!")
+            print(" Smallest possible shape (1,1,1) is a monomer - please check your input and try again")
+            exit
+
+        self.number_of_monomers = self.aggregate_shape[0] * self.aggregate_shape[1] * self.aggregate_shape[2]
+        # allow the user to specify the displacements along each axis (x, y, z) as a vector
+        if "displacement_vector" in args:
```
Comment (FoleyLab) [inline wptherml/exciton.py:91]: I also included the horizontal and vertical displacements into this displacement vector:
horizontal_displacement <-> displacement_vector[0]
vertical_displacement <-> displacement_vector[1]

and if one wants 3D aggreagates (corresponding to some aggregate_shape (Nx, Ny, Nz) with Nz >= 2)
then the displacement along the z axis is given by displacement_vector[2]

File: wptherml/exciton.py [hunk 6]
```diff
@@ -0,0 +1,594 @@
...
+
+        if "refractive_index" in args:
+            self.refractive_index = args["refractive_index"]
+        else:
+            self.refractive_index = 1
+
+
+        self.wvlngth_variable = np.arange(0, 400.001, 0.01)
+
+    def _compute_cartesian_coordinates(self):
```
Comment (FoleyLab) [inline wptherml/exciton.py:110]: Using the aggregate_shape and displacement_vector attributes, this method will fill out the cartesian coordinates for each monomer.

File: wptherml/exciton.py [hunk 7]
```diff
@@ -0,0 +1,594 @@
...
+            the distance separating the donor and acceptor
+        n : float
+            refractive index of the medium
+        """
+
+        r_scalar = np.sqrt(np.dot(r_vector, r_vector))
+    
+        return (1 / (self.refractive_index ** 2 * r_scalar ** 3)) * (np.dot(self.transition_dipole_moment, self.transition_dipole_moment) - 3 * (np.dot(self.transition_dipole_moment, r_vector) * np.dot(r_vector, self.transition_dipole_moment)) / r_scalar ** 2)
+
+    def build_exciton_hamiltonian(self):
```
Comment (FoleyLab) [inline wptherml/exciton.py:219]: With the cartesian coordinates specified, this will compute the Hamiltonian 1D, 2D, or 3D aggregates.

File: wptherml/exciton.py (hunk 1 above)
Comment (lanetolley) [reply wptherml/exciton.py]: Those are in atomic units! I took out the original displacement_between_monomers

File: examples/simple_dynamics.py [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+'displacement_between_monomers' : np.array([1, 0, 0]),
+'transition_dipole_moment' : np.array([0, 0, 0.0]),
+'refractive_index' : 1.0
+
+}
+
+dynamics_test = sf.spectrum_factory("Frenkel", dynamics_args)
+print(dynamics_test.exciton_hamiltonian)
+
+#def test_rk_exciton():
```
Comment (FoleyLab) [inline examples/simple_dynamics.py]: @lanetolley remove the commented out section of the example for tidiness

File: examples/simple_dynamics.py [hunk 9]
```diff
@@ -0,0 +1,59 @@
+# Import package, test suite, and other packages as needed
+import wptherml
+import numpy as np
+import pytest
+import sys
+
+
+""" perform a series of tests that would apply to the simple
```
Comment (FoleyLab) [inline examples/simple_dynamics.py]: @lanetolley This comment doesn't seem related to this example

File: wptherml/data/Pyrene.json [hunk 10]
```diff
@@ -0,0 +1,13 @@
...
+    "level_of_theory": "DFT / 6-31G**",
```
Comment (FoleyLab) [inline wptherml/data/Pyrene.json]: Which functional?  Is TDDFT used for the excited state?

File: wptherml/data/amine_TTZ_acid.json [hunk 11]
```diff
@@ -0,0 +1,13 @@
...
+    "level_of_theory": "DFT / 6-31G**",
```
Comment (FoleyLab) [inline wptherml/data/amine_TTZ_acid.json]: @lanetolley which functional?  Is TDDFT used for the S1 state?

File: wptherml/data/amine_TTz_fluoro.json [hunk 12]
```diff
@@ -0,0 +1,13 @@
...
+    "level_of_theory": "DFT / 6-31G**",
```
Comment (FoleyLab) [inline wptherml/data/amine_TTz_fluoro.json]: functional + tddft

File: wptherml/data/amine_TTz_ol.json [hunk 13]
```diff
@@ -0,0 +1,13 @@
...
+    "level_of_theory": "DFT / 6-31G**",
```
Comment (FoleyLab) [inline wptherml/data/amine_TTz_ol.json]: functional + tddft

File: wptherml/data/cyano_TTZ_acid.json [hunk 14]
```diff
@@ -0,0 +1,13 @@
...
+    "level_of_theory": "DFT / 6-31G**",
```
Comment (FoleyLab) [inline wptherml/data/cyano_TTZ_acid.json]: functional + tddft

File: wptherml/tests/test_exciton.py [hunk 15]
```diff
@@ -0,0 +1,117 @@
...
+"""
+
+# Import package, test suite, and other packages as needed
+import wptherml
+import numpy as np
+import pytest
+import sys
+
+
+""" perform a series of tests that would apply to the simple
```
Comment (FoleyLab) [inline wptherml/tests/test_exciton.py]: @lanetolley update this comment to be descriptive of the systems being tested

File: wptherml/tests/test_exciton.py [hunk 16]
```diff
@@ -0,0 +1,117 @@
...
+    _H_12 = exciton_test._compute_H0_element(1, 2)
+
+    _expected_H_11 = exciton_test.exciton_energy
+    _expected_H_12 = 0.
+
+    assert np.isclose(_H_11, _expected_H_11, 1e-5)
+    assert np.isclose(_H_12, _expected_H_12, 1e-5)
+
+
+def test_compute_dipole_dipole_coupling():
```
Comment (FoleyLab) [inline wptherml/tests/test_exciton.py:97]: @lanetolley This unit test needs to be improved, it only tests a trivial case for when the transition dipole moment is zero and the monomers are separated by a very large distance.  

You should write a test for several different cases that will make sure that the calculation works under a variety of displacements and for the different instances of a transition dipole moment that might arise.

For example, can you have a TDM with non-zero x, y, and z components and then have unit tests for when:

a. the monomers are displaced along x only but by a small amount
b. the monomers are displaced along y only but by a small amount
c. the monomers are displaced along z only but by a small amount
d. the monomers are displaced along x, y, and z but by a small amount

File: wptherml/tests/test_exciton.py [hunk 17]
```diff
@@ -0,0 +1,117 @@
...
+
+    _V_test = exciton_test._compute_dipole_dipole_coupling(0, 1)
+
+    _V_expected = 0
+
+    assert np.isclose(_V_test, _V_expected)
+
+
+
+def test_build_exciton_hamiltonian():
```
Comment (FoleyLab) [inline wptherml/tests/test_exciton.py:170]: @lanetolley This unit tests needs to be improved, it only tests a simple case where there are 2 monomers and no coupling.  Just like the dipole-dipole coupling, you need to test several non-trivial cases.

Since the dipole-dipole coupling unit test will catch any potential bugs in the dipole-dipole coupling function between any 2 monomers, the unit test for the build_exciton_hamiltonian method should focus on if the correct coupling is captured when there are more than 2 monomers... i.e. maybe a 3 monomer case testing coupling between units all in-plane and then a 4 monomer case testing out-of-plane coupling.

File: wptherml/tests/test_exciton.py [hunk 18]
```diff
@@ -0,0 +1,117 @@
...
+    E1 = dynamics_test.exciton_energy
+    E2 = dynamics_test.exciton_energy
+    dynamics_test.build_exciton_hamiltonian()
+
+
+    dt1 = 0.01
+
+    tf = 1
+
+    D_analytical = np.array([[np.cos(E1-E2) - ci * np.sin(E1-E2), 0 + 0j], [0 + 0j, 0 + 0j]])
```
Comment (FoleyLab) [inline wptherml/tests/test_exciton.py]: @lanetolley I'm not sure I understand this unit test, can we discuss?

**Review by FoleyLab — CHANGES_REQUESTED** [review]
@lanetolley I have added comments to the test and example files where changes are suggested or required before merging; please have a look and address when you have the chance and let me know if you have any questions!

**Review by FoleyLab — COMMENTED** [review]
This fixes the issues with CI
